### PR TITLE
KAFKA-6958: Overload KTable methods to allow to name operation name using the new Named class

### PR DESCRIPTION
**Pull request by fhussonnois**
Hi @mjsax @bbejeck

This is the 4th PR for the KIP-307.
NOTE :[PR 6411](https://github.com/apache/kafka/pull/6411) should be merged first

Thanks a lot for the review.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 1]
```diff
@@ -353,21 +601,70 @@
...
+
+    /**
+     * Convert this changelog stream to a {@link KStream}.
+     * <p>
+     * Note that this is a logical operation and only changes the "interpretation" of the stream, i.e., each record of
+     * this changelog stream is no longer treated as an updated record (cf. {@link KStream} vs {@code KTable}).
+     *
+     * @return a {@link KStream} that contains the same records as this {@code KTable}
+     */
+    KStream<K, V> toStream();
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:645]: We didn't specify this in the KIP but I think `toStream()` and `toStream(mapper)`  should also have overrides with`Named`

**Review by bbejeck — COMMENTED** [review]
Made a pass over the `KTable` and `KTableImpl`.  Overall looks good but IMHO I think we should add `KTable.toStream()`.  Also, there aren't tests for `KTable`.

**Comment by bbejeck** [comment]
ping @vvcephei and @ableegoldman for reviews

**Comment by bbejeck** [comment]
@fhussonnois please rebase this PR

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 1 above)
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:645]: I've updated the KIP and the PR to the two methods above

**Comment by bbejeck** [comment]
@fhussonnois with #6411 merged can you rebase this PR? With the bulk of the work done so far, I think we can get this PR and the last PR for KIP-307 merged relatively quickly.

**Comment by fhussonnois** [comment]
Hi @bbejeck, this PR is rebased and ready for review. Thanks

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 2]
```diff
@@ -259,6 +421,89 @@
...
+     * For each {@code KTable} update the provided {@link ValueMapper} is applied to the value of the updated record and
+     * computes a new value for it, resulting in an updated record for the result {@code KTable}.
+     * Thus, an input record {@code <K,V>} can be transformed into an output record {@code <K:V'>}.
+     * This is a stateless record-by-record operation.
+     * <p>
+     * The example below counts the number of token of the value string.
+     * <pre>{@code
+     * KTable<String, String> inputTable = builder.table("topic");
+     * KTable<String, Integer> outputTable = inputTable.mapValue(new ValueMapper<String, Integer> {
+     *     Integer apply(String value) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:461]: nit: ditto

I'll go ahead and say we should do so for all examples (when possible of course) in the Javadoc and I won't point out anymore from this point.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 3]
```diff
@@ -224,6 +350,42 @@
...
+     * (with possibly a new type) in the new {@code KTable}, with default serializers, deserializers, and state store.
+     * For each {@code KTable} update the provided {@link ValueMapper} is applied to the value of the updated record and
+     * computes a new value for it, resulting in an updated record for the result {@code KTable}.
+     * Thus, an input record {@code <K,V>} can be transformed into an output record {@code <K:V'>}.
+     * This is a stateless record-by-record operation.
+     * <p>
+     * The example below counts the number of token of the value string.
+     * <pre>{@code
+     * KTable<String, String> inputTable = builder.table("topic");
+     * KTable<String, Integer> outputTable = inputTable.mapValue(new ValueMapper<String, Integer> {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: nit: change `mapValue` to take a lambda vs. an anonymous `ValueMapper`

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 4]
```diff
@@ -259,6 +421,89 @@
...
+     * For each {@code KTable} update the provided {@link ValueMapperWithKey} is applied to the value of the update
+     * record and computes a new value for it, resulting in an updated record for the result {@code KTable}.
+     * Thus, an input record {@code <K,V>} can be transformed into an output record {@code <K:V'>}.
+     * This is a stateless record-by-record operation.
+     * <p>
+     * The example below counts the number of token of value and key strings.
+     * <pre>{@code
+     * KTable<String, String> inputTable = builder.table("topic");
+     * KTable<String, Integer> outputTable = inputTable.mapValue(new ValueMapperWithKey<String, String, Integer> {
+     *     Integer apply(String readOnlyKey, String value) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: nit: same here

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 5]
```diff
@@ -696,40 +1590,45 @@
      * <td>&lt;K1:ValueJoiner(A,b)&gt;</td>
      * </tr>
      * <tr>
-     * <td>&lt;K1:C&gt;</td>
-     * <td>&lt;K1:C&gt;</td>
+     * <td>&lt;K1:null&gt;</td>
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:1578]: why these changes?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 6]
```diff
@@ -762,7 +1665,7 @@
      * <td>&lt;K1:A&gt;</td>
      * <td></td>
      * <td></td>
-     * <td></td>
+     * <td>&lt;K1:ValueJoiner(A,null)&gt;</td>
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:1653]: same here

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 7]
```diff
@@ -387,28 +469,34 @@ public String queryableStoreName() {
...
+    @Override
+    public <K1> KStream<K1, V> toStream(final KeyValueMapper<? super K, ? super V, ? extends K1> mapper,
+                                        final Named named) {
+        return toStream(named).selectKey(mapper);
+    }
+
     @Override
     public KTable<K, V> suppress(final Suppressed<? super K> suppressed) {
         final String name;
+        final NamedInternal renamed;
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: A while back there was a discussion on how to handle naming `suppression` and it's a different case.  We only want to invoke the `builder.newProcessorName` in cases where the user has not supplied a name.  So the code for suppress should stay the same.  

Although we might lose context as to why this is different, I believe the Javadoc contains the explanation so we should be covered.

\cc @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 8]
```diff
@@ -387,28 +469,34 @@ public String queryableStoreName() {
...
+            name = renamed.orElseGenerateWithPrefix(builder, SUPPRESS_NAME);
         } else {
             throw new IllegalArgumentException("Custom subclasses of Suppressed are not supported.");
         }
 
         final SuppressedInternal<K> suppressedInternal = buildSuppress(suppressed, name);
 
-        final String storeName =
-            suppressedInternal.name() != null ? suppressedInternal.name() + "-store" : builder.newStoreName(SUPPRESS_NAME);
+        final String storeName = renamed.suffixWithOrElseGet("-store", builder, SUPPRESS_NAME, true);
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: same for the store

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java [hunk 9]
```diff
@@ -67,6 +67,10 @@ String suffixWithOrElseGet(final String suffix, final String other) {
     }
 
     String suffixWithOrElseGet(final String suffix, final InternalNameProvider provider, final String prefix) {
+        return suffixWithOrElseGet(suffix, provider, prefix, false);
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java]: Were these changes put in only for working with `suppress`? If so we can revert them.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressTopologyTest.java [hunk 10]
```diff
@@ -62,16 +62,16 @@
         "      --> myname\n" +
         "      <-- KSTREAM-SOURCE-0000000005\n" +
         "    Processor: myname (stores: [myname-store])\n" +
-        "      --> KTABLE-TOSTREAM-0000000006\n" +
+        "      --> KTABLE-TOSTREAM-0000000008\n" +
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressTopologyTest.java]: Same here, since we'll leave `suppress` as is, these test changes can be reverted.

\cc @vvcephei

**Review by bbejeck — COMMENTED** [review]
Thanks @fhussonnois! I just have minor comments, but otherwise, this looks good.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:645]: good catch!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: Yep. Thanks, @bbejeck !

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressTopologyTest.java (hunk 10 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressTopologyTest.java]: Yep, we should not change these tests. Thanks!

**Review by vvcephei — CHANGES_REQUESTED** [review]
Hey @fhussonnois , thanks for this!

Once you address Bill's comments, this will be fine by me.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java (hunk 9 above)
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java]: This modification has been reverted

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 8 above)
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: This modification has been reverted

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 5 above)
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:1578]: I think, these changes come from javadoc shift due to new methods added.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 6 above)
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:1653]: I think, these changes come from javadoc shift due to new methods added.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 2 above)
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:461]: the javadoc has been updated - there was also an issue on the method name (ie : mapValue to mapValues)

**Review by vvcephei — APPROVED** [review]
Thanks for the update!

**Review by bbejeck — APPROVED** [review]
Thanks @fhussonnois, LGTM.

**Comment by bbejeck** [comment]
Merged #6412 into trunk.

**Comment by bbejeck** [comment]
Thanks again @fhussonnois!
